pkg/registry/core/service/storage/storage.go

```diff
@@ -641,7 +642,7 @@ func patchAllocatedValues(after After, before Before) {
 }
 
 func needsClusterIP(svc *api.Service) bool {
```

Comment:
hmm, we have exactly teh same function in strategy https://github.com/kubernetes/kubernetes/blob/ed72ee264eaa51d02bc1dc031408c40bba5bed3f/pkg/registry/core/service/strategy.go#L296-L301 , I wonder if we have the same issue there

Reply:
/hold

Reply:
Yes, we shouldn't make the change here, maybe we should make the change at line 591.
```diff
if needsClusterIP(oldSvc) && needsClusterIP(newSvc) {
- if newSvc.Spec.ClusterIP == "" {
+	// If oldSvc service type is Headless Service,
+	// we won't use it to override newSvc Spec.ClusterIP field
+	if newSvc.Spec.ClusterIP == "" && oldSvc.Spec.ClusterIP != core.ClusterIPNone {
		newSvc.Spec.ClusterIP = oldSvc.Spec.ClusterIP
	}
	if len(newSvc.Spec.ClusterIPs) == 0 && len(oldSvc.Spec.ClusterIPs) > 0 {
		newSvc.Spec.ClusterIPs = oldSvc.Spec.ClusterIPs
	}
}
```
What do you think?

Reply:
first let's discuss the issue, is not clear is a real bug